### PR TITLE
Guard Empty TaskResult Properties Type Assign

### PR DESCRIPTION
**Pull request by JeromeJu**
# Changes
This commit guards the empty taskResult Properties check, which would otherwise assign default ResultTypeObject to TaskResults with empty Properties maps.

Thanks to @Yongxuanzhang this would help not to default ResultTypeObject to Properties that are set as map.

/kind misc

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [n/a] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [n/a] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [n/a] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
NONE
```

### CODE REVIEW

**Comment by dibyom** [comment]
@JeromeJu do we know why we'd have `Properties` initialized but with length 0? It seems unlikely that would happen from user created yamls...does this get adding during conversion or something?

**Comment by JeromeJu** [comment]
> @JeromeJu do we know why we'd have `Properties` initialized but with length 0? It seems unlikely that would happen from user created yamls...does this get adding during conversion or something?

Yes it was introduced with the conversion here https://github.com/tektoncd/pipeline/blob/a7424c4885dbaab56a7a31cb9f98508a6f7fc3fd/pkg/apis/pipeline/v1beta1/result_conversion.go#L29. Does it make sense to guard it here instead of restoring the empty map back to nil in conversion?

> (Also, there is a [coverage drop](https://ff66c8a2d671591bf353ae91346b982098825d18cf682eacac64325-apidata.googleusercontent.com/download/storage/v1/b/tekton-prow/o/pr-logs%2Fpull%2Ftektoncd_pipeline%2F6570%2Fpull-tekton-pipeline-go-coverage-df%2F1649370343304859648%2Fartifacts%2Fline-cov.html?jk=Ac_6HjLoo7QGevS_7j15GtKZlN7Xd9Xm-j-fIiAbu8Upd70aEbHZxojMZ3WxXzmEPk96SjoPd-9vekRz8vdRUNpWwr1kX9fNt7rxjGy7inb_soOOmFdB8t_vSu7GgJ8sum6fQXdJYyEJfyxSEsEIqcl6u5Cl02diBorM_YyFg-5amWXW3d2QPxznEmS_rKeZ4MvD99H4ZG8LpVmUHHLw9xTaBx8OeYqsAHZwsHyqDFjBeMzoaVtX2tsKLN0Ka_hERC8tYHaZkq-sUiJjelBAR7oz3CrcjhJsDkx5iNu4BKq2BxCeWFyuvvacbw1dxr4LRfbTHDlIY9I0ikzeBIMeDOgqRhKETPouimso92AaCnZj8PwwOHEHYI4UNJYkvOlPktYUvYZLwmW924o7uqEFu-elAgFkR320cECQB9rUN1sHJQcODaXr2ODo7y3v5QopwDRv2bQH9eRFx-9QrAcjVJCYpFc4mcP0agcxHj4_vQ0wA-WZQmFIP1xy_ce6FhZjawIxG-WxJiCIphtUYn1LugWa-nzn0Ur0E_zA6zAz__fudsTDTYpOMssIuzhtCmPDDAtShQVkPOX84wR6j9IbGBunSC_j6Z3gi6CgVRGLmCpKPOq97dxkQClK5WtK2hQG7fhGD9ko2MuYj9zYH3NR_NzUWpc6XWWWAa8wQv9liiTLSY9vM5BvDra9PZPQdBvcvUXY05N8uXGKV3A2fTT7IIlanWnxt-L3IUOTNbiJFnprqbtFk_ROHAhg1p5Z4C8PcplTOZCwbok4ftHNEULXUyL08ppQStnv8yxudk0wwQSDi7dC8W1Nlqakf35pMrYLqhE1XvBQnAk-8DtoUy8oA9M5BjaUgQNyMBQShSTCBbWbEpPlCCW6LeNRvempA8RXUZ-wl8VwkGJw26RTO3QMJLEopSBWkbX81f81yPNJpHUS5wZintMzajSrzl41dG7kNi4XV1aLgLUhxxGXjRfbiC19Vere-SHSQc9FHANdUIKU00qgrAKzqXeyj9VCVbxy3SS5VDQoPaqAct3i5W_BjrBlNyyGzm-b3yuhFkU7KVpTvw9iNzgxYuP0CU9K1Rhzl8XVgivQ0YU8prmoVaX0YDIAYRciB37PbUpYqcQhlT7dJXgBcCOs6eWt35pGxP_J3wCjHuhMCApM_j_IN8x_xSYFPfPIWcsy_xqHHU28KSiSVKmKBeAtjfEhK6M184VwotzW4IhHTndoM5e6bMUCk18Y5eXR2NIQ2MiXrsbECyScJVTEYZAqLVdXzDrDMAPQ1_B9wdWNuojewahcMfq5OMmI7ZO6dzgDWzaab4q262efpMEZYQI&isca=1#file0), can we add a test to fix? )

Sorry this was previously when i introduced a typo, I believe right now the coverage should be restored though I could not retrigger the coverage test with

/test pull-tekton-pipeline-go-coverage

File: pkg/apis/pipeline/v1/result_defaults.go [hunk 1]
```diff
@@ -21,7 +21,7 @@ func (tr *TaskResult) SetDefaults(context.Context) {
 		return
 	}
 	if tr.Type == "" {
-		if tr.Properties != nil {
+		if tr.Properties != nil && len(tr.Properties) != 0 {
```
Comment (lbernick) [inline pkg/apis/pipeline/v1/result_defaults.go:24]:
```suggestion
		if len(tr.Properties) != 0 {
```

Comment (lbernick) [reply pkg/apis/pipeline/v1/result_defaults.go:24]: https://google.github.io/styleguide/go/decisions#nil-slices

**Review by lbernick — APPROVED** [review]

File: pkg/apis/pipeline/v1/result_defaults.go (hunk 1 above)
Comment (JeromeJu) [reply pkg/apis/pipeline/v1/result_defaults.go:24]: Learned  ✏️ 📓,  thanks for the pointer Lee!

**Comment by Yongxuanzhang** [comment]
I think params properties also have the same logic, should we fix that as well? It could be a separate PR

**Comment by JeromeJu** [comment]
> I think params properties also have the same logic, should we fix that as well? It could be a separate PR

Thanks, mended at https://github.com/tektoncd/pipeline/pull/6578

**Comment by jerop** [comment]
> > @JeromeJu do we know why we'd have `Properties` initialized but with length 0? It seems unlikely that would happen from user created yamls...does this get adding during conversion or something?
> 
> Yes it was introduced with the conversion here (thanks to the pointer by @Yongxuanzhang )
> 
> https://github.com/tektoncd/pipeline/blob/a7424c4885dbaab56a7a31cb9f98508a6f7fc3fd/pkg/apis/pipeline/v1beta1/result_conversion.go#L29

what do you think about fixing the conversion instead?

/hold

**Comment by jerop** [comment]
That's interesting, could you please add that context to the commit message and PR description?

This makes me wonder, will we treat empty arrays as nil too and default to string instead of interpreting them as an empty array result?

**Comment by dibyom** [comment]
> it seems that the root of the problem is the conversion, maybe the fix should be in the conversion?

Agreed. I think we do want to support empty objects (and not cast them as strings). 

Couldn't we just conditionally create `properties` if the type is object? https://github.com/tektoncd/pipeline/blob/a7424c4885dbaab56a7a31cb9f98508a6f7fc3fd/pkg/apis/pipeline/v1beta1/result_conversion.go#L29C1-L33

What do you think @lbernick @Yongxuanzhang ?

**Comment by lbernick** [comment]
+1 on fixing the conversion logic.

Re: supporting empty arrays and allowing results to be empty:

---- Object Results -----
I believe the effect of this change is that if you declare the following Task:

```
spec:
  results:
  - name: result-1
     properties: {}
```
then the task is declaring that it will emit a string result, when prior to this commit the task would be declaring that it would emit an object result. I don't think this commit would treat an empty object result in TaskRun status as a string result, as this commit only affects the "TaskResult" struct used in Task result definitions, not the TaskRunResult struct used in TaskRun statuses.

I don't quite understand what TEP-0075 means when it says "must be possible for a result object to be empty"-- does this mean it can be declared with empty properties, or that not all properties need values? It also says we're "suggesting infering some things such as [that type: string is being used for empty properties](https://github.com/tektoncd/community/blob/main/teps/0075-object-param-and-result-types.md#declaring-object-results-and-params)" but I can't find an example of this.

Here's an example from [TEP-0075](https://github.com/tektoncd/community/blob/main/teps/0075-object-param-and-result-types.md#declaring-object-results-and-params) of declaring an object result in a Task. The Task specifies the properties that should be present in the result. (Note: I can't find any documentation on this! the closest thing I could find was [these docs](https://github.com/tektoncd/pipeline/blob/main/docs/pipelines.md#emitting-results-from-a-pipeline) on passing object results from one task into another task's params.)

I don't know what it would mean to declare an object result in a task with no properties-- what would be the point? My instinct is that treating a Task result definition with empty properties as a string result is correct, and also aligns with the principle of not treating nil maps differently than empty maps.

I'm not sure whether it's possible in practice for a TaskRun to emit an empty object result. It's [OK for a task to emit more keys in an object result than it declared](https://github.com/tektoncd/community/blob/main/teps/0075-object-param-and-result-types.md#matching-interfaces-and-extra-keys), and [object params must have values for all keys](https://github.com/tektoncd/community/blob/main/teps/0075-object-param-and-result-types.md#required-vs-optional-keys). However, I'm not sure if the same applies to object results. Again though-- this commit should only affect task result definitions.

We really need some better docs here IMO!

---- Array results -----

It doesn't look like this commit addresses array results at all, but it seems like there's some concern about it.

To declare a [Task array result](https://github.com/tektoncd/pipeline/blob/main/docs/tasks.md#emitting-array-results) you have to specify:
```
spec:
  results:
  - name: result-1
     type: array
```
To declare a Task array param, you would need to use:
```
spec:
  params:
  - name: param-1
     type: array
```
or
```
spec:
  params:
  - name: param-1
     type: array
     default:
     - foo
     - bar
```
The only nil/empty mixup we could make here would be treating an empty default array as a string param, which I don't think is a concern here.

In the TaskRun:

```
status:
  results:
  - name: result-1
     value:
     - foo
     - bar
```

We need to make sure that we're not treating
```
status:
  results:
  - name: result-1
     value: {}
```
as a string result-- but I don't think that's relevant to this commit.


TL;DR I think this commit is correct but we also need to fix conversion logic.

**Comment by Yongxuanzhang** [comment]
+1 to fix the conversion and thanks @lbernick for opening the PR to fix it. 

I think we need to reach an agreement on https://github.com/tektoncd/pipeline/issues/6605 to decide whether we want to merge this PR or not.

**Comment by JeromeJu** [comment]
Closed as - https://github.com/tektoncd/pipeline/pull/6606 has been merged
